File: jieun^,^/BOJ_14891번_톱니바퀴_Implementation.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class BOJ_14891번_톱니바퀴_Implementation {
    static char[][] gear = new char[4][8];

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        
        // 톱니 바퀴의 상태 입력 받기
        for (int i=0; i<4; i++){
            String Line = br.readLine();
            for (int j=0; j<8; j++){
                gear[i][j]=Line.charAt(j);
            }
        }

        int K = Integer.parseInt(br.readLine());    // 회전 횟수
        StringTokenizer st;
        for (int k=0; k<K; k++){
            st = new StringTokenizer(br.readLine());
            int g = Integer.parseInt(st.nextToken());   // 몇 번째 톱니바퀴를 회전시킬지
            int dir = Integer.parseInt(st.nextToken()); // 어느 방향으로 회전시킬지

            // 한번 회전시킬 때마다 한 톱니바퀴는 최대 한 번만 회전해야함
            // 그것을 체크하기 위한 boolean 배열
            boolean checked[] = new boolean[4]; 
            Arrays.fill(checked, false);
            
            // K번 회전
            rotation(gear, g-1, dir, checked);
        }

        //for (int i=0; i<4; i++){System.out.println(Arrays.toString(gear[i]));}

        System.out.println(cal(gear));

    }

    public static void turn(char[][] gear, int g, int dir){
        char[] tmp = Arrays.copyOf(gear[g], gear[g].length);

        if (dir==1){    // 시계 방향
            for(int i=0; i<8; i++){ 
                if (i==0){gear[g][i]=tmp[7];}
                else{gear[g][i]=tmp[i-1];}
            }
        }else if(dir==-1){  // 반시계 방향
            for(int i=0; i<8; i++){
                if(i==7){gear[g][i]=tmp[0];}
                else{gear[g][i]=tmp[i+1];}
            }
        }

    }

    public static void rotation(char[][] gear, int g, int dir, boolean[] checked){
        char[] before = Arrays.copyOf(gear[g], gear[g].length);
        
        turn(gear, g, dir);
        checked[g]=true;    // g번째 톱니바퀴 회전시켰다고 체크해두기

        // 인접 톱니바퀴 체크
        // 오른쪽 톱니바퀴 체크
        if (g!=3){
            if (before[2]!=gear[g+1][6]){
                if (checked[g+1]==false){
                    rotation(gear, g+1, -dir, checked); // 오른쪽 톱니바퀴를 반대 방향으로 회전
                }
            }
        }

        // 왼쪽 톱니바퀴 체크
        if (g!=0){
            if (before[6]!=gear[g-1][2]){
                if (checked[g-1]==false){
                    rotation(gear, g-1, -dir, checked);
                }
            }
        }
    }

    public static int cal(char[][] gear){
        int ans=0;
        if (gear[0][0]=='1'){ans += 1;}
        if (gear[1][0]=='1'){ans += 2;}
        if (gear[2][0]=='1'){ans += 4;}
        if (gear[3][0]=='1'){ans += 8;}
        
        return ans;
    }
}

File: jieun^,^/BOJ_14891번_톱니바퀴_Implementation.py
### 순수하게 해당 톱니바퀴만 회전시키는 함수 ###
def turn(gear, g, dir): 
    tmp = gear[g].copy()

    if dir==1:  # 시계 방향       
        for i in range(8):
            if i==0: gear[g][i]=tmp[7]
            else: gear[g][i]=tmp[i-1]

    if dir==-1:       # 반시계 방향
        for i in range(8):
            if i==7: gear[g][i]=tmp[0]
            else: gear[g][i]=tmp[i+1]


def rotation(gear, g, dir, checked):    # gear: 톱니바퀴 배열, g: 회전시킬 톱니바퀴 번호, dir: 시계 or 반시계, checked: 이번 순서에 이미 회전시켰는지 체크
    before = gear[g].copy()
    turn(gear, g, dir)
    checked[g]=True

    ### 인접 톱니바퀴 체크 ###
    # 오른쪽 톱니바퀴 체크
    if g!=3:    # 제일 오른쪽 톱니바퀴가 아닌 경우
        if before[2]!= gear[g+1][6]:   # 오른쪽 톱니바퀴와 만나는 부분이 N과 S임
            if checked[g+1]!=True:
                rotation(gear, g+1, -dir, checked)


    # 왼쪽 톱니바퀴 체크
    if g!=0:    # 제일 왼쪽 톱니바퀴가 아닌 경우
        if before[6]!= gear[g-1][2]:
            if checked[g-1]!=True:
                rotation(gear, g-1, -dir, checked)


def calGear(gear):
    # 12시 방향이 S극일 때만 점수 획득
    ans = 0
    if gear[0][0]=='1': ans += 1   
    if gear[1][0]=='1': ans += 2
    if gear[2][0]=='1': ans += 4
    if gear[3][0]=='1': ans += 8    

    return ans

def solution():
    ### 톱니 바퀴의 현재 상태 입력 받기 ###
    gear = []
    for g in range(4):
        gear.append(list(input()))  # N: '0', S: '1'

    ### K번 회전하기 ###
    K = int(input())
    for k in range(K):
        g, d = map(int, input().split())    # a: 회전할 바퀴(1~4), b: 회전 방향 (1: 시계, -1: 반시계)
        rotation(gear, g-1, d, [False, False, False, False])

    #print("회전 후 gear: ", gear)


    ### 회전 후 계산하기 ###
    return calGear(gear)

print(solution())
